File: ios/Yizhidao/Data/ImaExplanationStore.swift
import Foundation

struct ImaExplanationEntry: Codable {
    let title: String
    let scripture: String
    let answer: String
}

private struct ImaExplanationsFile: Codable {
    let version: Int
    let source: String
    let entries: [String: ImaExplanationEntry]
}

enum ImaExplanationId {
    static func guaci(number: Int) -> String { String(format: "%02d-guaci", number) }
    static func tuanci(number: Int) -> String { String(format: "%02d-tuanci", number) }
    static func daxiang(number: Int) -> String { String(format: "%02d-daxiang", number) }

    /// `position` 为 1…6（初爻=1）
    static func yaoPair(number: Int, position: Int) -> String {
        String(format: "%02d-yao-%d", number, position - 1)
    }

    static func yong(number: Int) -> String { String(format: "%02d-yong", number) }
    static func wenyan(number: Int) -> String { String(format: "%02d-wenyan", number) }
}

struct ImaExplanationSelection: Identifiable {
    let id: String
    let entry: ImaExplanationEntry
}

@Observable
final class ImaExplanationStore {
    static let shared = ImaExplanationStore()

    private var entries: [String: ImaExplanationEntry] = [:]
    private(set) var source = ""

    init(bundle: Bundle = .main) {
        load(from: bundle)
    }

    func explanation(for id: String) -> ImaExplanationEntry? {
        entries[id]
    }

    func hasExplanation(for id: String) -> Bool {
        entries[id] != nil
    }

    private func load(from bundle: Bundle) {
        guard let url = bundle.url(forResource: "ImaExplanations", withExtension: "json") else {
            return
        }
        do {
            let data = try Data(contentsOf: url)
            let file = try JSONDecoder().decode(ImaExplanationsFile.self, from: data)
            source = file.source
            entries = file.entries
        } catch {
            assertionFailure("Failed to load ImaExplanations.json: \(error)")
        }
    }
}

File: android/app/src/main/java/com/yizhidao/app/ima/ImaExplanationStore.kt
package com.yizhidao.app.ima

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class ImaExplanationEntry(
    val title: String,
    val scripture: String,
    val answer: String,
)

@Serializable
private data class ImaExplanationsFile(
    val version: Int = 1,
    val source: String = "",
    val entries: Map<String, ImaExplanationEntry> = emptyMap(),
)

object ImaExplanationId {
    fun guaci(number: Int): String = "%02d-guaci".format(number)
    fun tuanci(number: Int): String = "%02d-tuanci".format(number)
    fun daxiang(number: Int): String = "%02d-daxiang".format(number)

    /** `position` 为 1…6（初爻=1） */
    fun yaoPair(number: Int, position: Int): String =
        "%02d-yao-%d".format(number, position - 1)
}

class ImaExplanationStore(jsonText: String) {
    private val file: ImaExplanationsFile

    val source: String
        get() = file.source

    init {
        val json = Json { ignoreUnknownKeys = true }
        file = json.decodeFromString(jsonText)
    }

    fun explanation(id: String): ImaExplanationEntry? = file.entries[id]

    fun hasExplanation(id: String): Boolean = file.entries.containsKey(id)
}
